Feature/ppm 5 handle smart steering bml requests and fix code review comments

adam1985d opened the pull request:
This PR handles comments raised in the previous controller DB PR - related to the runtime DB in the controller.
As part of this PR - I have also added handling for the following BML requests which complete the BML<->controller integration for the smart-steering feature.
ACTION_BML_CLIENT_GET_CLIENT_LIST_REQUEST
ACTION_BML_CLIENT_SET_CLIENT_REQUEST
ACTION_BML_CLIENT_GET_CLIENT_REQUEST

The above-mentioned Fixes are limited to the BML, BML-CLI, and controller.

The controller<->BPL integration is handled separately by the following PR:  #1502

adam1985d committed:
controller: db: change aging check to use the remaining time

The current condition of comparing the client's passed-timelife-delay
against max-timelife-delay is correct but doesn't reflect the actual
logic behind it which is "is remaining timelife over".
This change enables a more coherent comparison of
current-client-remaining-timelife-delay against a
candidate-client-for-removal-timelife-delay (if the DB is full).

Changed the aging condition to "client-remaining-time < 0".

PPM-5.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: db: change client_db_entry_to_mac to get a copy

The current client_db_entry_to_mac receives a const& to the mac address
and internally makes a copy to perform an in-place replacement of
characters.

Changed the function to receive a copy instead of a const& and changed
the implementation accordingly.

PPM-5.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: db: add `using ValuesMap` for cleaner code

Added `using ValuesMap = std::unordered_map<std::string, std::string>`
and replaced DB APIs and private function to use the new `ValuesMap`
instead of `std::unordered_map<std::string, std::string>`.

The above change is a fix to comment raised in a previous PR, and it
results in much cleaner code.

PPM-5.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: db: load_persistent_db_clients: handle db is full

As part of the load_persistent_db_clients() a corner-case where the user
added to the persistent DB more than the allowed max-clients-db-size
must be handled.

Added a check of the number-of-added-clients against the
max-clients-db-size - if the condition is met - the clients' DB is full.
If clients DB is full, find a candidate client for removal and compare
it against the client read from persistent DB and we are trying to add.

If the new client's remaining timelife is greater than the timelife of
the candidate_for_removal, then the candidate is removed from the
runtime DB and the persistent DB and the new client is added instead to
the runtime DB.

If the new client's remaining timelife is smaller or equal to the
timelife of the candidate_for_removal, then the new client is removed
from the persistent DB and we continue to the next client.

PPM-5.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: db: add API to get all configured clients

Preparative commit.

Added API to get all the clients in the runtime DB that have persistent
information configured.

The assumption is that a client with persistent data configured must
have a timestamp configured (which differs from the default value that
the not-configured clients have).

The API returns a dqueu of the MAC addresses of the above-mentioned
clients.

PPM-5.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: db: change selected_bands type

The current implementation of selected bands mistakenly used the
existing beerocks::eFreqType as the storage type. This is not extensible
and doesn't define the full possible configurations.
Moreover, by using the newly introduced eClientSelectedBands instead,
which is used in the BML<->controller messaging, we can simplify the set
and get operations and any future extension of supported values.

Changed in node class the storage type of the selected bands and its
description.
Changed the DB APIs to use the int8_t instead of beerocks::eFreqType.
Changed the selected bands` default value from
beerocks::eFreqType::FREQ_UNKNOWN to beerocks::PARAMETER_NOT_CONFIGURED.

PPM-5.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: read list of clients from DB

Changed the ACTION_BML_CLIENT_GET_CLIENT_LIST_REQUEST handling to read
the clients from persistent DB using the
get_clients_with_persistent_data_configured() DB API.

Removed the related cpp-check issue from cppcheck_existing_issues.txt.

PPM-5.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: read client information from DB

Change the ACTION_BML_CLIENT_GET_CLIENT_REQUEST handling to read the
client information from the DB using the DB APIs.
The handling also takes care of the following scenarios:
"client does not exist in DB" - handled as an error.
"client timestamp not exist" - handled as an error.

PPM-5.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: set client information to DB

Change the ACTION_BML_CLIENT_SET_CLIENT_REQUEST handling to configure
the client information to the runtime DB using the DB APIs.
If persistent DB is enabled - the update_client_persistent_db() API
called to update the persistent DB with the client information (to
prevent multiple persistent DB access overhead).

Note that if the client is not yet part of the runtime DB, a node for
it added before-hand.
Moreover, managing the persistent DB max size limit is handled
internally and is transparent to the user.

PPM-5.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
bml: BML_CLIENT: change sta_mac type from char to uint8_t

The type of the mac address parameter is incorrect.

Changed sta_mac from char[6] to uint8_t[6].

PPM-5.

Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>
Signed-off-by: Adam Dov <adamx.dov@intel.com>

itayx reviewed (approved):
Looks great

adam1985d committed:
beerocks_cli_bml: Fix client APIs

- In client_set_client changed incoming type from INT_ARG to STRING_ARG.
To use parameters as optional parameters, we need to retrieve them from
the CLI as strings and convert them to integers.

- Added int conversion in the client_set_client LOG print.
To print the incoming values in client_set_client, need to convert them
to type int, otherwise, they will be printed as Char.

- Added tlvf::mac_to_string in the client_get_client LOG print.
The received mac address needs to be converted to a readable format to
be printed.

PPM-5.

Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com
Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: bml: fix return values

Changed returning error values in the code to a negative values as
expected by the BML.

PPM-5.

Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>
Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: bml: refactor client_get_client

Added verification of "is someone waiting for the promise" before
handling the response.
Implement the function with error-handling first according to the
agreed coding guidelines.
Fixed broken return-on-error in the "get_client" implementation.

PPM-5.

Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>
Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: bml: validate set_client input parameters

The BML APIs provide an optional-parameters interface, meaning the user
can use the API with some of the parameters set to "NOT_CONFIGURED"
value, and these parameters not configured in the DB.

However, using the API with all the parameters set to "NOT_CONFIGURED"
value is not a valid case and should be handled as an error.

Added verification that at least one of the given parameters not set to
"NOT_CONFIGURED" value.

PPM-5.

Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>
Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
db: set the initial-radio when stay-on-initial-radio is set

When configuring the stay_on_initial_radio the initial_radio needs to
be set as well.
If stay_on_initial_radio is disabled - clear the initial_radio
configuration.
If stay_on_initial_radio is enabled and the client is connected - set
the initial_radio to the parent radio (not bssid) mac.

PPM-5.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: db: set the initial-radio on client connection

The stay-on-initial-radio can be configured for disconnected clients as
well as clients which never connected yet.
This means, that when the client connects, we need to check if the
stay-on-initial-radio is enabled and if so set the initial-radio.
Note that the initial-radio is set only if it wasn't set before - to
prevent overriding previous configuration.
The value set to the initial-radio is the parent radio (not bssid) mac.

Added the above-described functionality to the
association_handling_task::finalize_new_connection() which is called on
client connection.

PPM-5.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
bml: remove the unsupported stay_on_selected_device

The introduction of the stay_on_selected_device configuration, its DB
support, and utilization delayed to the next phase of the
smart-steering feature.

Removed the stay_on_selected_device from the BML interface - from
set_client and get_client APIs related structs.
Removed the stay_on_selected_device from the BML CLI interface - both
from set_client and get_client CLI functions.

Note that the stay_on_selected_device infrastructure in the BML message
to/from the controller is not removed, but not used.

PPM-5.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
bml: cli: fix selected_bands debug prints

Removed excess spaces from selected-bands debug print.

PPM-5.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

kantera800 commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -759,20 +762,20 @@ class db {
      * @param mac MAC address of a client.
-     * @return Selected band/bands.
+     * @return Selected band/bands. Possible values are bitwise options of eClientSelectedBands.
      */
-    beerocks::eFreqType get_client_selected_bands(const sMacAddr &mac);
+    int8_t get_client_selected_bands(const sMacAddr &mac);
```
why not use std::bitmap ?

adam1985d replied:
I think you meant `std::bitset` - so will answer to that.

I could and probably should - thanks.
Will introduce this change as part of the next PR.

morantr commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -2270,13 +2270,13 @@ int cli_bml::client_get_client(const std::string &sta_mac)
                 return "Disabled";
             if (val & BML_CLIENT_SELECTED_BANDS_24G)
-                ret += "2.4 Ghz,";
+                ret += "2.4Ghz,";
```
With space, it looks better IMO. I also think it is the correct way to write it.
BTW, it should be `GHz` and not `Ghz`. (capital H).

adam1985d replied:
accepted partially.
changed `Ghz` to `GHz` - but the spacing makes it much less readable so removed the spaces.

morantr commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1965,6 +1965,12 @@ int bml_internal::client_set_client(const sMacAddr &sta_mac, const BML_CLIENT_CO
+        LOG(WARNING) << "No parameter is requested to be configured, returning";
+        return (-BML_RET_INVALID_ARGS);
+    }
+
```
Why do it here and not on `cli_bml::client_set_client`?
BTW, `client_set_client` is a terrible name, I don't even understand what it does.

adam1985d replied:
1. moving the verification to the `cli_bml::client_set_client` means that any application that links to the bml.h and uses the bml_client_set_client() API directly will not be verified.

morantr commented on controller/src/beerocks/master/db/node.h:
```diff
@@ -292,8 +292,9 @@ class node {
-    beerocks::eFreqType client_selected_bands = beerocks::eFreqType::FREQ_UNKNOWN;
+    // Default value is PARAMETER_NOT_CONFIGURED.
+    // Possible values are bitwise options of eClientSelectedBands.
+    int8_t client_selected_bands = beerocks::PARAMETER_NOT_CONFIGURED;
```
Then why didn't you set its type to `eClientSelectedBands`?  😕

adam1985d replied:
to support "parameter not configured value"

I added a suggestion to out "fixups and improvements" PR to extend the eClientSelectedBands to have a "not configured" value - but since this change has a tree-wide effect - wanted to introduce it (id decided to) as hotfix commit (and of course change the db to hold eClientSelectedBands accordingly.

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2808,9 +2808,10 @@ bool db::load_persistent_db_clients()
+    int add_node_error_count                        = 0;
+    int set_node_error_count                        = 0;
+    int clients_added_no_error                      = 0;
+    int clients_not_added_or_removed_due_to_full_db = 0;
```
Why all of the are signed? Can they have a negative value?

adam1985d replied:
you are right.
changed them to `uint16_t` (to support changes in the max_clients_in_persistent_db to larger values without the need to update the code, as well as user pre-defined persistent DB configuration with a number of clients greater than the allowed number of clients in persistent DB)

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2897,6 +2901,61 @@ bool db::load_persistent_db_clients()
             continue;
         }
 
+        // If clients DB is full - find candidate for removal and compare against the current client.
+        // Note that this is a corner case and at the init stage where this functionality is performed we do
+        // not expect for this condition to be met.
+        // This is only for robustness against user misuse (adding manually more clients than clients_persistent_db_max_size).
```
This all operation is a lot of code, might be a good idea to insert it to a function.

adam1985d replied:
true - will be encapsulated as part of the next PR
Looking through it and other functions in the DB I think there might be a more generic approach to this flow that the add operation handles the "if I don't have room in DB" and not the load function directly.

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2980,6 +2980,24 @@ bool db::load_persistent_db_clients()
+    LOG_IF(configured_clients.empty(), DEBUG) << "No clients are found";
+
+    return configured_clients;
+}
```
`std::vector` is the worst container to pick for this operation, since you push elements and causing reallocation.
Please use `std::deque`.

adam1985d replied:
replaced with deque (as well as dependant following commits)

morantr commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -2060,9 +2060,8 @@ void son_management::handle_bml_message(Socket *sd,
-        // TODO: replace with a list of configured clients read from controller-db
-        std::vector<sMacAddr> client_list;
-        if (!client_list.size()) {
+        std::vector<sMacAddr> client_list = database.get_clients_with_persistent_data_configured();
```
`auto client_list = ...` (nitpick).

adam1985d replied:
fixed, was actually a remainder from previous PR that only declared it without really using it (that's why it was explicit and not `auto`)

morantr commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -2110,10 +2110,99 @@ void son_management::handle_bml_message(Socket *sd,
+            (request->client_config().selected_bands == PARAMETER_NOT_CONFIGURED)) {
+            LOG(ERROR)
+                << "Received ACTION_BML_CLIENT_SET_CLIENT request without parameters to configure";
+            send_response(false);
```
Missing `break`.

adam1985d replied:
added the missing `break;` - thanks

morantr commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -2110,10 +2110,99 @@ void son_management::handle_bml_message(Socket *sd,
+            LOG(DEBUG) << "Setting a client which doesn't exist in DB, adding client to DB";
+            if (!database.add_node(client_mac)) {
+                LOG(ERROR) << "Failed to add client node for client " << client_mac;
+                send_response(false); //Fail.
```
Ther comment is redundant, please remove it.

adam1985d replied:
removed the comment

morantr commented on controller/src/beerocks/master/tasks/association_handling_task.cpp:
```diff
@@ -372,6 +372,18 @@ void association_handling_task::finalize_new_connection()
+                    LOG(WARNING) << "Failed to set client " << client_mac << "  initial radio to "
+                                 << parent_radio_mac;
+                }
+            }
```
I think it will be a good idea to add a comment about what you are doing here, and why.

adam1985d replied:
added a comment describing the reason we set the initial-radio here and what is the pre-condition to setting it.

adam1985d committed:
controller: db: fix function return value on success to true

As part of testing the full flows of the persistent Db we encountered
an issue where a successful removal of a client entry from the
persistent DB resulted in a false-positive error.
The cause of this error is that remove_client_entry_and_update_counter()
in the db class returned false both for success and for failure.

Changed the successful return value of the above-mentioned function
from false to true.

Signed-off-by: Adam Dov <adamx.dov@intel.com>
